File: src/main/java/com/github/morningzeng/toolset/model/Children.java
package com.github.morningzeng.toolset.model;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.github.morningzeng.toolset.Constants.IconC;
import com.google.common.collect.Lists;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.FieldNameConstants;
import lombok.experimental.SuperBuilder;

import javax.swing.Icon;
import java.util.List;
import java.util.Objects;

/**
 * @author ch.zeng
 * @since 2023-03-30
 */
@Getter
@Setter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
@FieldNameConstants(innerTypeName = "Children_")
public abstract class Children<T> {

    @Builder.Default
    private final boolean directory = false;
    /**
     * children
     */
    @JsonAlias({"item"})
    private List<T> children;
    @JsonIgnore
    private T parent;

    public void addChild(T child) {
        if (Objects.isNull(this.children)) {
            this.children = Lists.newArrayList();
        }
        this.children.add(child);
    }

    public Icon icon() {
        return this.isGroup() ? IconC.FOLDER_COLOR : IconC.TREE_NODE;
    }

    public abstract String name();

    public boolean isGroup() {
        return this.directory;
    }

}

File: src/main/kotlin/com/github/morningzeng/toolbox/model/Children.kt
package com.github.morningzeng.toolbox.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.github.morningzeng.toolbox.Constants.IconC
import javax.swing.Icon

/**
 * @author Morning Zeng
 * @since 2025-05-19
 */
abstract class Children<T : Children<T>>(
    var children: MutableList<T> = mutableListOf(),
    @JsonIgnore
    var parent: T? = null,
    var sorted: Int = 0,
    var directory: Boolean = false,
) {
    companion object {
        fun <T : Children<T>> comparable(): Comparator<T?> =
            Comparator.comparing<T, Int> { it.sorted }.thenComparing { it.name() }
    }

    fun addChild(child: T) {
        @Suppress("UNCHECKED_CAST")
        child.parent = this as T
        children.add(child)
    }

    abstract fun name(): String

    fun icon(): Icon {
        return if (directory) IconC.FOLDER_COLOR else IconC.TREE_NODE
    }
}
